### PR TITLE
Improve duration calc

### PR DESCRIPTION
**Pull request by ritzalam**
The calculation of user duration assumed that number of joins should equal number of lefts. If they are not, set the duration to zero.

Sometimes, a user rejoins multiple times because of connection issues.

2020-10-26 08:07:17 +0000 cur_duration=00:00:00 duration=00:00:00 join=join
2020-10-26 08:07:58 +0000 cur_duration=00:00:41 duration=00:00:41 join=join
2020-10-26 08:08:25 +0000 cur_duration=00:00:27 duration=00:01:08 join=left
2020-10-26 08:08:50 +0000 cur_duration=00:00:00 duration=00:01:08 join=join
2020-10-26 08:09:35 +0000 cur_duration=00:00:45 duration=00:01:53 join=join
2020-10-26 08:09:55 +0000 cur_duration=00:00:20 duration=00:02:13 join=left
2020-10-26 08:10:25 +0000 cur_duration=00:00:00 duration=00:02:13 join=join
2020-10-26 08:10:45 +0000 cur_duration=00:00:20 duration=00:02:33 join=join
2020-10-26 08:10:45 +0000 cur_duration=00:00:00 duration=00:02:33 join=left
2020-10-26 08:11:14 +0000 cur_duration=00:00:00 duration=00:02:33 join=join
2020-10-26 08:11:45 +0000 cur_duration=00:00:31 duration=00:03:04 join=left
2020-10-26 08:12:12 +0000 cur_duration=00:00:00 duration=00:03:04 join=join
2020-10-26 08:21:54 +0000 cur_duration=00:09:42 duration=00:12:46 join=join
2020-10-26 08:53:05 +0000 cur_duration=00:31:11 duration=00:43:57 join=left
2020-10-26 08:56:31 +0000 cur_duration=00:03:26 duration=00:47:23 join=left

Try to calculate user duration when user rejoins the meeting several times.

### CODE REVIEW

File: calc-duration.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+
```
Comment (kepstin) [inline calc-duration.rb]: I think you left this file in the PR by accident?

Most of the contents here would be better as tests.

File: lib/bbbevents/recording.rb [hunk 2]
```diff
@@ -58,6 +59,7 @@ def initialize(events_xml)
       @attendees.values.each do |att|
         att.leaves << @finish if att.joins.length > att.leaves.length
         att.duration = total_duration(att)
+        att.duration_human = Time.at(att.duration).utc.strftime("%H:%M:%S")
```
Comment (kepstin) [inline lib/bbbevents/recording.rb]: Using `Time.at` (which is actually intended for absolute times relative to the epoch) for duration formatting is problematic due to time zone issues. (edit: not an issue here, i see you're using `.utc`) This wouldn't work at all for durations >24h, too.

File: lib/bbbevents/recording.rb [hunk 3]
```diff
@@ -146,13 +148,44 @@ def timestamp_conversion(base)
 
     # Calculates an attendee's duration.
     def total_duration(att)
-      return 0 unless att.joins.length == att.leaves.length
-      total = 0
+      joins_leaves_arr = []
+      att.joins.each { |j| joins_leaves_arr.append({:time => j.to_i, :datetime => j, :event => :join})}
+      att.leaves.each { |j| joins_leaves_arr.append({:time => j.to_i, :datetime => j, :event => :left})}
+
+      #puts "BEGIN ============"
```
Comment (kepstin) [inline lib/bbbevents/recording.rb]: Please remove the commented out/dead code.

**Review by kepstin — CHANGES_REQUESTED** [review]
This really needs some tests added.

File: lib/bbbevents/recording.rb (hunk 2 above)
Comment (kepstin) [reply lib/bbbevents/recording.rb]: Come to think of it, doing human date formatting in bbb-events is out of scope, that should be done by whatever tool is consuming the bbb-events data and displaying it to the user.

File: calc-duration.rb (hunk 1 above)
Comment (ritzalam) [reply calc-duration.rb]: Removed.

File: lib/bbbevents/recording.rb (hunk 2 above)
Comment (ritzalam) [reply lib/bbbevents/recording.rb]: Removed.

File: lib/bbbevents/recording.rb (hunk 3 above)
Comment (ritzalam) [reply lib/bbbevents/recording.rb]: Removed.

**Review by kepstin — APPROVED** [review]

**Pull request merged**